akka-projection-core/src/test/scala/akka/projection/ProjectionBehaviorSpec.scala

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.Promise
+
+import akka.Done
+import akka.NotUsed
+import akka.actor.ClassicActorSystemProvider
+import akka.actor.testkit.typed.scaladsl.LogCapturing
+import akka.actor.testkit.typed.scaladsl.ScalaTestWithActorTestKit
+import akka.actor.testkit.typed.scaladsl.TestProbe
+import akka.stream.KillSwitches
+import akka.stream.scaladsl.Sink
+import akka.stream.scaladsl.Source
+import org.scalatest.wordspec.AnyWordSpecLike
+
+class ProjectionBehaviorSpec extends ScalaTestWithActorTestKit with AnyWordSpecLike with LogCapturing {
+
+  "A ProjectionBehavior" must {
+
+    "start immediately on demand" in {
+
+      val testProbe = testKit.createTestProbe[ProbeMessage]()
+      val src = Source(1 to 2)
+      testKit.spawn(ProjectionBehavior(() => TestProjection(src, testProbe)))
+
+      testProbe.expectMessage(StartObserved)
+      testProbe.expectMessage(Consumed(1, "1"))
+      testProbe.expectMessage(Consumed(2, "1-2"))
+      // good, things are flowing
+
+    }
+
+    "stop actor when receiving stop message" in {
+
+      val testProbe = testKit.createTestProbe[ProbeMessage]()
+      val src = Source(1 to 2)
+      val projectionRef = testKit.spawn(ProjectionBehavior(() => TestProjection(src, testProbe)))
+
+      testProbe.expectMessage(StartObserved)
+      testProbe.expectMessage(Consumed(1, "1"))
+      testProbe.expectMessage(Consumed(2, "1-2"))
+      // good, things are flowing
+
+      projectionRef ! ProjectionBehavior.Stop
+      testProbe.expectMessage(StopObserved)
+
+      testProbe.expectTerminated(projectionRef)
+
+    }
+
+    "failures stopping stream don't block stopping the actor" in {
+
+      val testProbe = testKit.createTestProbe[ProbeMessage]()
+      val src = Source(1 to 2)
+      val projectionRef = testKit.spawn(ProjectionBehavior(() => TestProjection(src, testProbe, failToStop = true)))
+
+      testProbe.expectMessage(StartObserved)
+      testProbe.expectMessage(Consumed(1, "1"))
+      testProbe.expectMessage(Consumed(2, "1-2"))
+      // good, things are flowing
+
+      // this Stop will crash the actor
+      projectionRef ! ProjectionBehavior.Stop
+
+      testProbe.expectTerminated(projectionRef)
+
+    }
+  }
+
+  sealed trait ProbeMessage
+  case object StartObserved extends ProbeMessage
+  case class Consumed(n: Int, currentState: String) extends ProbeMessage
+  case object StopObserved extends ProbeMessage
+
+  /*
+   * This TestProjection has a internal state that we can use to prove that on restart,
+   * the actor is taking a new projection instance.
+   */
+  case class TestProjection(src: Source[Int, NotUsed], testProbe: TestProbe[ProbeMessage], failToStop: Boolean = false)
```

Comment:
I think it's better to define such test classes in `ProjectionBehaviorSpec` companion object. More realistic and avoids accidentally using something that is in scope from the test class.